src/mrack/providers/provider.py

```diff
@@ -30,16 +37,116 @@ def name(self):
 
     async def validate_hosts(self, hosts):
         """Validate that host requirements are well specified."""
-        return
+        raise NotImplementedError()
 
     async def can_provision(self, hosts):
         """Check that provider has enough resources to provison hosts."""
         raise NotImplementedError()
 
-    async def provision_hosts(self, hosts):
-        """Provision hosts and wait for it to finish."""
+    async def create_server(self, req):
+        """Request and create resource on selected provider."""
+        raise NotImplementedError()
+
+    async def wait_till_provisioned(self, resource):
+        """Wait till resource is provisioned."""
+        raise NotImplementedError()
+
+    async def provision_hosts(self, hosts, logger=None):
+        """Provision hosts based on list of host requirements.
+
+        Main provider method for provisioning.
+
+        First it validates that host requirements are valid and that
+        provider has enough resources(quota).
+
+        Then issues provisioning and waits for it succeed. Raises exception if any of
+        the servers was not successfully provisioned. If that happens it issues deletion
+        of all already provisioned resources.
+
+        Return list of information about provisioned servers.
+        """
+        if not logger:
+            logger = logging.getLogger(self.name)
+
+        logger.info("Validating hosts definitions")
+        await self.validate_hosts(hosts)
+        logger.info("Host definitions valid")
+
+        logger.info("Checking available resources")
+        can = await self.can_provision(hosts)
+        if not can:
+            raise ValidationError("Not enough resources to provision")
+        logger.info("Resource availability: OK")
+
+        started = datetime.now()
+
+        count = len(hosts)
+        logger.info(f"Issuing provisioning of {count} hosts")
+        create_servers = []
+        for req in hosts:
+            awaitable = self.create_server(req)
+            create_servers.append(awaitable)
+        create_resps = await asyncio.gather(*create_servers)
+        logger.info("Provisioning issued")
+
+        logger.info("Waiting for all hosts to be available")
+        wait_servers = []
+        for create_resp in create_resps:
+            awaitable = self.wait_till_provisioned(create_resp)
+            wait_servers.append(awaitable)
+
+        server_results = await asyncio.gather(*wait_servers)
+        provisioned = datetime.now()
+        provi_duration = provisioned - started
+
+        logger.info("All hosts reached provisioning final state (ACTIVE or ERROR)")
+        logger.info(f"Provisioning duration: {provi_duration}")
+
+        errors = [res for res in server_results if res["status"] == "ERROR"]
```

Comment:
This status field and it's values are OpenStack specific. Part of `server_result` object can be unified between providers.

Reply:
Got it i will try to update this asap

Reply:
I have added parse_errors method.

Reply:
Final solution is "no-matter-the-provider-proof", in the future we can solve the way we handle the status everywhere.